driver-core/src/main/com/mongodb/internal/time/Timer.java

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.mongodb.internal.time;
+
+import com.mongodb.annotations.Immutable;
+import com.mongodb.internal.VisibleForTesting;
+
+import java.util.concurrent.TimeUnit;
+
+import static com.mongodb.assertions.Assertions.assertTrue;
+import static com.mongodb.internal.VisibleForTesting.AccessModifier.PRIVATE;
+import static java.util.concurrent.TimeUnit.NANOSECONDS;
+
+/**
+ * A <a href="https://docs.oracle.com/javase/8/docs/api/java/lang/doc-files/ValueBased.html">value-based</a> class
+ * for tracking elapsed time. The maximum duration this class can measure reliably is {@link Long#MAX_VALUE} nanoseconds,
+ * which is more than 292 years. The class must not be used to measure longer durations.
+ * <p>
+ * This class is not part of the public API and may be removed or changed at any time.</p>
+ */
+@Immutable
+public final class Timer {
+    private static final Timer USELESS = new Timer(0);
```

Comment:
This naming seems unclear.

Reply:
`USELESS` is what will likely be called a default value when [`null`-restricted types](https://openjdk.org/jeps/401) are implemented. Values of these types may have a heap layout that is even more optimal than that of [value objects](https://openjdk.org/jeps/8277163) (essentially, on par with the layout in C/C++/Rust/etc.).

However, since `USELESS` is confusing, I will use `null` as `Timeout.timer` in situations when the value of this field is not needed. I am postponing the change until the concerns in https://github.com/mongodb/mongo-java-driver/pull/1164#discussion_r1289051261 are resolved.

Reply:
I refactored `Timer` into `TimePoint`, and started using `null` instead of `Timer.useless`.